### PR TITLE
[MOB-2565] Browserstack Upload Automation via Fastlane

### PR DESCRIPTION
**Pull request by d4r1091**
[MOB-2565]

## Context

As part of the Firefox Upgrade work, we would like to automate the upload of the newest builds so that QA won't need to manually checkout branches and build to produce a `.app` to upload via Web to the Browserstack portal.

## Approach

- Add the Fastlane's browserstack plugin
- implement into Fastlane as part of the AppCenter work as the resulting build is the same
- Env vars securely stored in CircleCI

## Other

- Updated Fastlane
- Once we'll get it done (Firefox upgrade merged), the target branch will be always `main` 

### Checklist

- [x] I verified the workflow correctly triggered

[MOB-2565]: https://ecosia.atlassian.net/browse/MOB-2565?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -40,7 +40,7 @@ jobs:
           name: Build and deploy to Testflight
           command: bundle exec fastlane testflight_beta
 
-  build-and-deploy-appcenter:
+  build-and-deploy-appcenter-and-browserstack:
```
Comment (lucaschifino) [inline .circleci/config.yml:43]: Personally I'd say this is more organised if split into two separate jobs that run in the same workflow consecutively or in parallel

File: fastlane/Fastfile [hunk 2]
```diff
@@ -38,8 +38,8 @@ platform :ios do
     setup_circle_ci
   end
 
-  desc "Deploy to MS App-Center"
-  lane :appcenter do
+  desc "Deploy to MS App-Center + Browserstack"
+  lane :upload_to_appcenter_and_browserstack do
```
Comment (lucaschifino) [inline fastlane/Fastfile:42]: Or maybe at least could be two different fastlane actions to avoid this `and` pattern and leaving a single responsibility.

**Review by lucaschifino — APPROVED** [review]
Nice! 🤖 

Added a couple comments about organising them into separate jobs, let me know what you think 🙏 

Already approving it anyway as it works both ways and maybe I'm missing a reason to split them 🙂

**Comment by lucaschifino** [comment]
> @lucaschifino will comment here. Good call on splitting them, personally would have gone that way as default. However, for the sake of saving CircleCI credits as well as the timing hence servers energy (♻️), I went after coupling the job. Whenever an action like `gym` performs a build that results in a successful one, some lane-specific environment variables are being set. Those will contain the path to the resulting Derived Data as well as the Product (.app, .ipa) paths. By having a single `gym` action, we'll get for free the resulting product needed to upload the build to those two different services without having to perform another process or call another lane. There is a middle step like creating another lane but that lane would be called as part of the appcenter one in order to be able to pass the parameters to its subsequent separated browserstack one. However, that would result in fake decoupling as both will be inevitably linked anyway.

Got it, makes sense! Thanks for explaining! ❤️♻️ 

Follow up question: if we leave them as consecutive actions on the same CircleCI workflow (with the one that generates the .ipa first) can't we still separate them but use the same previously generated file? I understand the file path is an environment variable right now, wonder wether we could set it up for the whole workflow instead of specific jobs.

FYI: If that is possible but takes too much effort, also no problem leaving it like you did imo.

**Comment by d4r1091** [comment]
Yeah can do that 💯 - not so much as too much effort no worries

**Comment by d4r1091** [comment]
@lucaschifino achieved what we agreed on by utilizing a bit of hard coding but still coherent and unified. What we intended was indeed tried by myself in previous experiences and forgot not being possible.

File: .circleci/config.yml [hunk 3]
```diff
@@ -64,8 +64,14 @@ jobs:
...
-          name: Build and deploy to AppCenter
-          command: bundle exec fastlane appcenter
+          name: Build the Ecosia app
+          command: bundle exec fastlane build_ecosia_app
+      - run:
+          name: Upload to AppCenter
+          command: bundle exec fastlane upload_to_appcenter
+      - run:
+          name: Upload to Browserstack
+          command: bundle exec fastlane upload_to_browserstack
```
Comment (lucaschifino) [inline .circleci/config.yml:74]: 👏

**Review by lucaschifino — APPROVED** [review]
Looks great and already more organised, thanks! 🎉

**Pull request merged**